### ISSUE TITLE
Download statistics plugin

### ISSUE THREAD

**Issue by mathstuf** [girder/girder#1989]
A plugin to track download statistics of files would be nice to have.

I think the best would be to track counts for downloads started (byte 0 sent), downloads completed (final byte sent), and maybe total requests.

Cc: @utkarshayachit @jourdain

**Comment by mgrauer** [comment]
Girder provides an SFTP server in addition to the web-app server, and currently the SFTP server doesn't allow for plugins, meaning we wouldn't be able to track download statistics through a plugin via the SFTP server in its current form.

Download statistics should be tracked in a way that is stateless on the web-app server, so that it would work with multiple app servers running at the same time.

**Comment by zachmullen** [comment]
Eventually I think the SFTP server should support plugins. We'd probably need to mock out the `apiRoot` field during `load`, but otherwise it should work.

**Comment by sutartmelson** [comment]
Is the idea to have the individual stats attached to each item? Or stored to be viewed collectively in the plugin control panel, for example. 

@jeffbaumes

**Comment by zachmullen** [comment]
For now let's keep them per-file and use the `increment` method on the file model to atomically increase the download count.

**Comment by sutartmelson** [comment]
I'm struggling to figure out a way to track downloads completed. Downloads started, just checking when byte 0 is requested, is straightforward. Total requests is also straightforward. I'm not sure if there's a way to check when the last byte was requested?

**Comment by zachmullen** [comment]
Perhaps in the generator where we stream the file, we can trigger an event after the loop terminates, assuming the last byte was requested. That should only get triggered if the server actually sent the last byte, not if the request was canceled prior to that.

**Comment by manthey** [comment]
I think you'd have to indicate this within each assetstore type, and couldn't do it generally.  For the filesystem assetstore, if `endByte` in the `downloadFile` method is equal to `file['size']`, and if `readLen` is 0 in the `stream` generator, then we could trigger an event marking that the entire download has been sent.

**Comment by zachmullen** [comment]
I think we can do it generally, but we'd have to introduce a wrapper generator inside the download method in the file model.

**Comment by sutartmelson** [comment]
If a client requests a file download(starting at byte zero, no end byte), is it possible for that request to turn into multiple requests? Or is counting the total requests for when a client specifically requests a partial file download?

**Comment by manthey** [comment]
The client can ask to download a file and have the connection fail or get aborted part way through.  It then could ask to resume the download using a range request for finish the download (and that can get interrupted, too).  For instance, I might curl a download, loose my connection, and ask to resume the download with curl's `--continue-at -` option.

**Comment by sutartmelson** [comment]
Okay I think I follow now. 

So wrapping the generator simply like this in the file model could work, right?
```python
    def download(self, file, offset=0, headers=True, endByte=None,
                 contentDisposition=None, extraParameters=None):
        """
        Use the appropriate assetstore adapter for whatever assetstore the
        file is stored in, and call downloadFile on it. If the file is a link
        file rather than a file in an assetstore, we redirect to it.

        :param file: The file to download.
        :param offset: The start byte within the file.
        :type offset: int
        :param headers: Whether to set headers (i.e. is this an HTTP request
            for a single file, or something else).
        :type headers: bool
        :param endByte: Final byte to download. If ``None``, downloads to the
            end of the file.
        :type endByte: int or None
        :param contentDisposition: Content-Disposition response header
            disposition-type value.
        :type contentDisposition: str or None
        :type extraParameters: str or None
        """
        yield self._download(file, offset=offset, headers=headers, endByte=endByte,
                             contentDisposition=contentDisposition,
                             extraParameters=extraParameters)
        # Trigger download complete

    def _download(self, file, offset=0, headers=True, endByte=None,
                 contentDisposition=None, extraParameters=None):
        if file.get('assetstoreId'):
            return self.getAssetstoreAdapter(file).downloadFile(
                file, offset=offset, headers=headers, endByte=endByte,
                contentDisposition=contentDisposition,
                extraParameters=extraParameters)
        elif file.get('linkUrl'):
            if headers:
                raise cherrypy.HTTPRedirect(file['linkUrl'])
            else:
                endByte = endByte or len(file['linkUrl'])

                def stream():
                    yield file['linkUrl'][offset:endByte]
                return stream
        else:  # pragma: no cover
            raise Exception('File has no known download mechanism.')
```

**Comment by zachmullen** [comment]
We could actually just modify the existing download method in the File model. Rather than naively return the value from the assetstore adapter, we could inspect it, check if it's a generator or function, and if it is, and if the final byte of the file is being requested, then we wrap the return value in our own generator that triggers the `download.complete` event at the end.

**Comment by sutartmelson** [comment]
Hm. I thought `download` was already required to be a generator.

**Comment by zachmullen** [comment]
There are two behaviors I know of: returning a generator, and raising an exception that will cause an HTTP redirect. So, for our current set of assetstores, it would be safe to blindly assume that we are dealing with a generator, although it wouldn't defend us against future use cases we may not have thought of.

**Comment by zachmullen** [comment]
My suggestion still won't handle cases like the S3 assetstore that redirect to an outside service. We don't really have a way of knowing for sure whether a user completely downloaded a file from S3, or canceled that connection partway through.

**Comment by sutartmelson** [comment]
Oh okay. So more along these lines? (Still not sure how to actually trigger an event. Are we creating a new event for this?)

```python
    def download(self, file, offset=0, headers=True, endByte=None,
                 contentDisposition=None, extraParameters=None):
        """
        Use the appropriate assetstore adapter for whatever assetstore the
        file is stored in, and call downloadFile on it. If the file is a link
        file rather than a file in an assetstore, we redirect to it.

        :param file: The file to download.
        :param offset: The start byte within the file.
        :type offset: int
        :param headers: Whether to set headers (i.e. is this an HTTP request
            for a single file, or something else).
        :type headers: bool
        :param endByte: Final byte to download. If ``None``, downloads to the
            end of the file.
        :type endByte: int or None
        :param contentDisposition: Content-Disposition response header
            disposition-type value.
        :type contentDisposition: str or None
        :type extraParameters: str or None
        """
        if file.get('assetstoreId'):
            fileDownload = self.getAssetstoreAdapter(file).downloadFile(
                file, offset=offset, headers=headers, endByte=endByte,
                contentDisposition=contentDisposition,
                extraParameters=extraParameters)
            if isinstance(fileDownload, six.types.GeneratorType):
                def wrap():
                    yield fileDownload
                    # Trigger download complete event
                return wrap
            else:
                return fileDownload
        elif file.get('linkUrl'):
            if headers:
                raise cherrypy.HTTPRedirect(file['linkUrl'])
            else:
                endByte = endByte or len(file['linkUrl'])

                def stream():
                    yield file['linkUrl'][offset:endByte]
                return stream
        else:  # pragma: no cover
            raise Exception('File has no known download mechanism.')
```

**Comment by zachmullen** [comment]
Something like that, but I don't know if you can wrap generators like that, I think you have to iterate over them and yield each chunk.

Yes, this would be a new event, I suggest calling it 'download.complete'.

**Comment by brianhelba** [comment]
Fixed by #2103
